packages/grafana-data/src/transformations/transformDataFrame.ts

```diff
@@ -1,37 +1,62 @@
+import { EMPTY, Observable, of } from 'rxjs';
+import { concatMap, expand, map, mergeMap, reduce } from 'rxjs/operators';
+
 import { DataFrame, DataTransformerConfig } from '../types';
 import { standardTransformersRegistry } from './standardTransformersRegistry';
 
+const getTransform = (config: DataTransformerConfig, data: Observable<DataFrame[]>): Observable<DataFrame[]> => {
+  const info = standardTransformersRegistry.get(config.id);
+
+  if (!info) {
+    return data;
+  }
+
+  return data.pipe(
+    mergeMap(before => {
+      const defaultOptions = info.transformation.defaultOptions ?? {};
+      const options = { ...defaultOptions, ...config.options };
+
+      return info.transformation.transformer(options, before).pipe(
+        map(after => {
+          if (after === before) {
+            return after;
+          }
+
+          // Add a key to the metadata if the data changed
+          for (const series of after) {
+            if (!series.meta) {
+              series.meta = {};
+            }
+
+            if (!series.meta.transformations) {
+              series.meta.transformations = [info.id];
+            } else {
+              series.meta.transformations = [...series.meta.transformations, info.id];
+            }
+          }
+
+          return after;
+        })
+      );
+    })
+  );
+};
+
 /**
  * Apply configured transformations to the input data
  */
-export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): DataFrame[] {
-  let processed = data;
-  for (const config of options) {
-    const info = standardTransformersRegistry.get(config.id);
-
-    if (!info) {
-      return data;
-    }
-
-    const defaultOptions = info.transformation.defaultOptions ?? {};
-    const options = { ...defaultOptions, ...config.options };
-    const transformer = info.transformation.transformer(options);
-    const after = transformer(processed);
-
-    // Add a key to the metadata if the data changed
-    if (after && after !== processed) {
-      for (const series of after) {
-        if (!series.meta) {
-          series.meta = {};
-        }
-        if (!series.meta.transformations) {
-          series.meta.transformations = [info.id];
-        } else {
-          series.meta.transformations = [...series.meta.transformations, info.id];
-        }
+export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): Observable<DataFrame[]> {
+  return of(of(data)).pipe(
```

Comment:
What is `of(of(data))`?

Reply:
Good question, it's a stream of streams or Observable of Observable or a Higher Order Observable. Let me know if it's unclear.

Reply:
An easier example here: https://stackblitz.com/edit/pg2o7z?devtoolsheight=33&file=index.ts

Reply:
:exploding_head: I'll believe you :)  

just spent 10mins in docs and yes -- i think i get it!

Reply:
I struggled to create an easier RxJs implementation of transformDataFrame, but it has a unique set of challenges:

- Each output from the stream uses the previous output as it's input
- We need to iterate over all supplied configs
- We're only interested in the last result

I'll have Another look and see if I can make it more readable

Reply:
@ryantxu I made the `transformDataFrame` more readable, let me know if it's better, thank you.

Reply:
yes better -- thanks